be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/DbActor.scala

```diff
@@ -168,84 +180,111 @@ object DbActor extends AskPatternConstants {
+    private def write(channel: Channel, message: Message): DbActorMessage = {
+      val channelDb: DB = channelsMap.get(channel) match {
+        case Some(channelDb) => channelDb
+        case _ => createDatabase(channel)
+      }
 
-        val channelDb: DB = channelsMap.get(channel) match {
-          case Some(channelDb) => channelDb
-          case _ => createDatabase(channel)
-        }
+      val messageId: Hash = message.message_id
+      Try(channelDb.put(messageId.getBytes, message.toJsonString.getBytes)) match {
+        case Success(_) =>
+          log.debug(s"Actor $self (db) wrote message_id '$messageId' on channel '$channel'")
+          DbActorWriteAck
+        case Failure(exception) =>
+          log.debug(s"Actor $self (db) encountered a problem while writing message_id '$messageId' on channel '$channel'")
+          DbActorNAck(ErrorCodes.SERVER_ERROR.id, exception.getMessage)
+      }
+    }
 
-        val messageId: Hash = message.message_id
-        Try(channelDb.put(messageId.getBytes, message.toJsonString.getBytes)) match {
-          case Success(_) =>
-            log.info(s"Actor $self (db) wrote message_id '$messageId' on channel '$channel'")
-            sender ! DbActorWriteAck
-          case Failure(exception) =>
-            log.info(s"Actor $self (db) encountered a problem while writing message_id '$messageId' on channel '$channel'")
-            sender ! DbActorNAck(ErrorCodes.SERVER_ERROR.id, exception.getMessage)
+    private def read(channel: Channel, messageId: Hash): DbActorMessage = {
+      if (channelsMap.contains(channel)) {
+        val channelDb: DB = channelsMap(channel)
+        Try(channelDb.get(messageId.getBytes)) match {
+          case Success(bytes) if bytes != null =>
+            DbActorReadAck(Some(Message.buildFromJson(new String(bytes, StandardCharsets.UTF_8))))
+          case _ =>
+            DbActorNAck(ErrorCodes.SERVER_ERROR.id, "Unknown read database error")
         }
+      } else {
+        DbActorReadAck(None)
+      }
+    }
 
-      case Read(channel, messageId) =>
-        log.info(s"Actor $self (db) received a READ request for message_id '$messageId' from channel '$channel'")
-
-        if (channelsMap.contains(channel)) {
-          val channelDb: DB = channelsMap(channel)
-          Try(channelDb.get(messageId.getBytes)) match {
-            case Success(bytes) if bytes != null =>
-              sender ! DbActorReadAck(Some(Message.buildFromJson(new String(bytes, StandardCharsets.UTF_8))))
-            case _ =>
-              sender ! DbActorNAck(ErrorCodes.SERVER_ERROR.id, "Unknown read database error")
-          }
+    private def catchup(channel: Channel): DbActorMessage = {
+      @scala.annotation.tailrec
+      def buildCatchupList(iterator: DBIterator, acc: List[Message]): List[Message] = {
+        if (iterator.hasNext) {
+          val value: Message = Message.buildFromJson(new String(iterator.next().getValue, StandardCharsets.UTF_8))
```

Comment:
there's no shame in introducing intermediate variables 😅 it might be worth it to setup the string first, then only build the message